rs/crypto/src/vetkd/mod.rs

```
@@ -0,0 +1,560 @@
+        );
+        self.metrics.observe_duration_seconds(
+            MetricsDomain::VetKd,
+            MetricsScope::Full,
+            "create_encrypted_key_share",
+            MetricsResult::from(&result),
+            start_time,
+        );
+        debug!(logger;
+            crypto.description => "end",
+            crypto.is_ok => result.is_ok(),
+            crypto.error => log_err(result.as_ref().err()),
+            crypto.vetkd_key_share => log_ok_content(&result),
+        );
+        result
+    }
+
+    fn verify_encrypted_key_share(
+        &self,
+        signer: NodeId,
+        key_share: &VetKdEncryptedKeyShare,
+        args: &VetKdArgs,
+    ) -> Result<(), VetKdKeyShareVerificationError> {
+        let log_id = get_log_id(&self.logger);
+        let logger = new_logger!(&self.logger;
+            crypto.log_id => log_id,
+            crypto.trait_name => "VetKdProtocol",
+            crypto.method_name => "verify_encrypted_key_share",
+        );
+        debug!(logger;
+            crypto.description => "start",
+            crypto.signer => format!("{}", signer),
+            crypto.vetkd_key_share => format!("{}", key_share),
+        );
+        let start_time = self.metrics.now();
+        let result = verify_encrypted_key_share_internal(
+            &self.lockable_threshold_sig_data_store,
+            self.registry_client.as_ref(),
+            &self.csp,
+            key_share,
+            signer,
+            args,
+        );
+        self.metrics.observe_duration_seconds(
+            MetricsDomain::VetKd,
+            MetricsScope::Full,
+            "verify_encrypted_key_share",
+            MetricsResult::from(&result),
+            start_time,
+        );
+        debug!(logger;
+            crypto.description => "end",
+            crypto.is_ok => result.is_ok(),
+            crypto.error => log_err(result.as_ref().err()),
+        );
+        result
+    }
+
+    fn combine_encrypted_key_shares(
```

In the [ticket description](https://dfinity.atlassian.net/browse/CRP-2599) I found the following:

> ASSUMPTION: Consensus calls this method only once it has received SUFFICIENT (i.e., reconstruction-threshold many) shares, whereby the individual shares MIGHT be invalid (in the sense that that node-signature is valid, but the encrypted share is not valid).

I believe this is not currently the case, as we leave determining the correct reconstruction-threshold up to crypto, and call the function as soon as we have _some_ shares for the same request.

---

Thanks for raising this!

This assumption was put there because my understanding is that Consensus actually does know the reconstruction threshold (it must know it, because Crypto knows the reconstruction threshold from Consensus). Do I understand correctly that at the place where you call `combine_encrypted_key_shares` you don't have this threshold easily available?

In any case, calling `combine_encrypted_key_shares` is fine and safe even with too little shares. We have a check in our crypto library that will eventually lead to a `VetKdKeyShareCombinationError::UnsatisfiedReconstructionThreshold`. The only downside I see with calls to `combine_encrypted_key_shares` with too few shares is unnecessary use of resources: concretely, we load/clone all data for the respective NI-DKG-ID from the threshold sig data store, deserialize the (G2) master public key, deserialize the (G1) encryption/transport public key, and deserialize plus (unfortunately) clone all shares we get.

What we could do is have an additional/redundant check of the reconstruction threshold early on in `combine_encrypted_key_shares` to fail fast and do as little unnecessary work as possible.

---

Something like [this](https://github.com/dfinity/ic/pull/3565/commits/a894eb8e506cf676cfc8edbf1c95df80269bf3ee) (cc @randombit).

---

Ah, you're right. The NiDkgId is part of the VetKD arguments, so we could look up the corresponding transcript in the latest summary block, which contains the threshold. I guess the effort to do this would be similar to the function you implemented above.

I see [similar functions](https://sourcegraph.com/github.com/dfinity/ic/-/blob/rs/crypto/src/sign/canister_threshold_sig/schnorr.rs?L189) also exist for ECDSA and Schnorr, but those could be implemented on consensus side as well, if needed.

---

Let's leave the [added](https://github.com/dfinity/ic/pull/3565/commits/a894eb8e506cf676cfc8edbf1c95df80269bf3ee) check then also for vetKD to be consistent with ECDSA/Schnorr, at least for now.